Split auto and interactive logic

freeekanayaka opened the pull request:
This set of changes splits the auto and interactive logic of the "lxd init" command into separate methods, as anticipated a while ago. It's a purely refactoring branch and there is no logic change, except that now the auto and interactive modes will have the same rollback-on-failure semantics as preseed (something I'm assuming is desirable).

It also makes both auto and interactive code leverage the exact same implementation as the preseed mode (essentially the cmdInit.apply() method), so we have a single logical path for all of them and keep things DRY.

The code to ask questions and to populate the cmdInitData structure that gets passed to apply() by the auto and interactive modes has also been split to standalone methods, for easier reading of top-level code and maintenance/understandability of individual askXXX and fillXXX logic.

Finally a few more unit tests and an integration test for the interactive mode have been added to increase confidence against regression.

The commit messages explain the various steps of the refactoring, but it's probably easier to have a look at the final main_init.go file directly to have a feel of the new code, which is hopefully an improvement over the current form.

brauner commented:
Fails on all build because of some go formatting issues, I think. :)

freeekanayaka commented:
@brauner lights are green now, yes there was a failure of static analysis plus another couple issues that I fixed.

brauner commented on lxd/main_init.go:
```diff
@@ -293,14 +293,23 @@ they otherwise would.
-			}
-		}
+		data.Config["core.https_address"] = ""
+		data.Config["core.trust_password"] = ""
```
I'm pretty sure this can be kicked out.

brauner replied:
I really fail to see why this is only done when we setup storage.

freeekanayaka replied:
Me neither, I dropped it.

brauner commented on lxd/main_init.go:
```diff
@@ -118,179 +105,32 @@ func (cmd *CmdInit) runAutoOrInteractive(c lxd.ContainerServer, backendsAvailabl
+		}
 
 		// FIXME: Allow to configure multiple storage pools on auto init
 		// run if explicit arguments to do so are passed.
```
Are we already allowing this now?

freeekanayaka replied:
No, because I don't think there's currently a way to express that with the current --auto command line flags (you can only specify the configuration for a single storage).

However considering that the --preseed option supports that, and it should eventually supersede --auto (which I guess could be dropped in LXD version 3), perhaps I can just drop this FIXME comment.

brauner replied:
Yeah.

freeekanayaka replied:
@brauner: dropped.

freeekanayaka commented:
jenkins: test this please

brauner commented:
Hm, I'm confused by

```
error: cannot listen on https socket: listen tcp: address -1: invalid port
```

when the testsuite runs
```
lxd init --storage-backend zfs --storage-create-loop 1 --storage-pool lxdtest-USU-init --auto
```

can you look into this, please?

freeekanayaka commented:
@brauner so re the Jenkins failure you point to, it seems that in the end the reset of core.https_address was actually doing "something", in the sense that if storage was passed via --auto, but no network, then the init command would be happy. However not resetting it ends up in the failure you saw and that I can reproduce locally:

```
root@x1:/tmp/lxd# LXD_DIR=$(pwd) /home/free/go/bin/lxd init --storage-backend zfs --storage-create-loop 1 --storage-pool test-init --auto
rror: cannot listen on https socket: listen tcp: invalid port -1
```

Also note that there is small bug in 2.13 an earlier, producing the same error if you don't specify a port:

```
root@x1:/tmp/lxd# LXD_DIR=$(pwd) /home/free/go/bin/lxd init --auto --network-address=127.0.0.1
error: cannot listen on https socket: listen tcp: invalid port -1
```

I fixed the issue by making --auto only tweak the network if --network-address is given, and have --network-port default to 8443 if it's not given.

I also fixed the flakiness in unit tests trying to use the hard-coded 9999 port, and modified them to allocate an unused port instead.

brauner commented:
One more question about this before merging:

> @brauner so re the Jenkins failure you point to, it seems that in the end the reset of core.https_address was actually doing "something", in the sense that if storage was passed via --auto, but no network, then the 
init command would be happy. However not resetting it ends up in the failure you saw and that I can reproduce locally:

So the question that remains to be answered is whether users can use `--auto` to only setup storage without setting up network without running into this problem? If not, then we should handle this.

freeekanayaka commented:
@brauner yes, with the fix I made (amended to the last commit) --auto can be used correctly to set storage only without network, network only without storage (which will default to the dir backend) or both. This fix also solves the existing problem that I mentioned above, existing in the previous implementation (iow what's in 2.13 and earlier), where you would get a runtime error if passing --network-address without --network-port. Now not passing --network-port will just use the default value.

brauner commented:
@freeekanayaka, you need to rebase. :) @stgraber made the necessary change to initialize the correct default port. The change should be kept by this branch. :)

freeekanayaka commented:
@brauner I've rebased the branch, note that the fix was already present in this branch, so I basically just resolved the conflict (just reverting @stgraber's change since it was the same as in this branch but in a different location in the file).

For reference, the line that contains that fix is:

https://github.com/lxc/lxd/pull/3328/files#diff-cdcf75ac4ef08c96178e6dddbfde242dR115
